can/can.cpp

```diff
@@ -0,0 +1,255 @@
+#ifndef CONFIG_DISABLE_MODULE_CAN
+
+#include "can.hpp"
+
+namespace halex {
+
+Can::Can(CAN_HandleTypeDef *hcan) : hcan(hcan) {
+
+}
+
+CAN_StatusType Can::init() {
+    if (HAL_CAN_Init(hcan) != HAL_OK) {
+        return CAN_StatusType::CAN_Fail_Init;
+    }
+    return static_cast<CAN_StatusType>(HAL_CAN_Start(hcan));
+}
+
+void Can::setFilterActivationState(uint32_t state) {
+    filterConfig.FilterActivation = state;
+}
+
+void Can::setFilterMode(CAN_FilterMode filterMode) {
+    this->filterMode = filterMode;
+
+    switch (this->filterMode) {
+        case CAN_FilterMode::PATH_ONE_TYPE_STD_OR_EXT_ID_GROUP:
+            filterConfig.FilterMode = CAN_FILTERMODE_IDMASK;
+            filterConfig.FilterScale = CAN_FILTERSCALE_32BIT;
+            break;
+
+        case CAN_FilterMode::PATH_TWO_TYPE_STD_OR_EXT_ID:
+            filterConfig.FilterMode = CAN_FILTERMODE_IDLIST;
+            filterConfig.FilterScale = CAN_FILTERSCALE_32BIT;
+            break;
+
+        case CAN_FilterMode::PATH_TWO_TYPE_STD_ID_GROUP:
+            filterConfig.FilterMode = CAN_FILTERMODE_IDMASK;
+            filterConfig.FilterScale = CAN_FILTERSCALE_16BIT;
+            break;
+
+        case CAN_FilterMode::PATH_FOUR_TYPE_STD_ID:
+
+            filterConfig.FilterMode = CAN_FILTERMODE_IDLIST;
+            filterConfig.FilterScale = CAN_FILTERSCALE_16BIT;
+            break;
+
+        case CAN_FilterMode::PATH_ALL_ID:
+            filterConfig.FilterMode = CAN_FILTERMODE_IDMASK;
+            filterConfig.FilterScale = CAN_FILTERSCALE_32BIT;
+            filterConfig.FilterIdHigh = 0x0000;
+            filterConfig.FilterIdLow = 0x0000;
+            filterConfig.FilterMaskIdHigh = 0x0000;
+            filterConfig.FilterMaskIdLow = 0x0000;
+            break;
+    }
+}
+
+void Can::setFilterBank(uint32_t CAN2_filterBankNumber) {
+    filterConfig.FilterBank = 0;
+    filterConfig.SlaveStartFilterBank = CAN2_filterBankNumber;
+}
+
+void Can::setStoreRxFifo(uint32_t rxFifo) {
+    filterConfig.FilterFIFOAssignment = rxFifo;
+}
+
+CAN_ClassSettingStatus Can::setFourTypePathId(uint32_t id1, uint32_t id2, uint32_t id3, uint32_t id4) {
+    if ((filterMode != CAN_FilterMode::PATH_ALL_ID) && (filterMode != CAN_FilterMode::PATH_FOUR_TYPE_STD_ID)) {
+
+        return CAN_ClassSettingStatus::WRONG_SET_PATH_ID_FUNCTION_USED;
+    } else if (filterMode == CAN_FilterMode::PATH_ALL_ID) {
+
+        return CAN_ClassSettingStatus::ALL_ID_PATH;
+    }
+
+    filterId[0] = id1;
+    filterId[1] = id2;
+    filterId[2] = id3;
+    filterId[3] = id4;
+
+    filterConfig.FilterIdHigh = filterId[0] << 5;
+    filterConfig.FilterIdLow = filterId[1] << 5;
+    filterConfig.FilterMaskIdHigh = filterId[2] << 5;
+    filterConfig.FilterMaskIdLow = filterId[3] << 5;
+
+    return CAN_ClassSettingStatus::NON_ERROR;
+}
```

Comment:
```suggestion
CAN_ClassSettingStatus Can::setFourTypePathId(uint32_t idHigh, uint32_t idLow, uint32_t maskIdHigh, uint32_t maskHighLow) {
    if ((filterMode != CAN_FilterMode::PATH_ALL_ID) && (filterMode != CAN_FilterMode::PATH_FOUR_TYPE_STD_ID)) {
        return CAN_ClassSettingStatus::WRONG_SET_PATH_ID_FUNCTION_USED;
    } else if (filterMode == CAN_FilterMode::PATH_ALL_ID) {
        return CAN_ClassSettingStatus::ALL_ID_PATH;
    }

    filterConfig.FilterIdHigh = idHigh << 5;
    filterConfig.FilterIdLow = idLow << 5;
    filterConfig.FilterMaskIdHigh = maskIdHigh << 5;
    filterConfig.FilterMaskIdLow = maskHighLow << 5;

    return CAN_ClassSettingStatus::NON_ERROR;
}
```

[SHOULD] filterConfig の名前に合わせたいです。また、filterId を使う必要はないです

Reply:
f29a400b6ba36b73d43ebd92af531b01314fad33 で対応